torchtune/utils/_distributed.py

```diff
@@ -345,12 +348,45 @@ def get_full_model_state_dict(
     """
     sharded_sd = model.state_dict()
     cpu_state_dict = {}
-    for param_name, sharded_param in sharded_sd.items():
-        full_param = sharded_param.full_tensor()
-        if is_rank_zero:
-            cpu_state_dict[param_name] = full_param.cpu()
-        else:
-            del full_param
+    has_nf4 = any(
+        isinstance(param._local_tensor, NF4Tensor) for param in model.parameters()
+    )
+    if has_nf4:
+        from torch.distributed._composable.fsdp.fully_shard import FSDPModule
+
+        # Iterating from lowerer modules to higher
+        # Unsharding lora adapters before unsharding transformer block
+        for module_name, module in reversed(list(model.named_modules())):
+            if not isinstance(module, FSDPModule):
+                continue
+            module.unshard(async_op=False)
+            if is_rank_zero:
+                module_name = module_name.replace(f".{_CHECKPOINT_WRAPPED_MODULE}", "")
+                for local_fqn, param in module.named_parameters():
+                    local_fqn = local_fqn.replace(f".{_CHECKPOINT_WRAPPED_MODULE}", "")
```

Comment:
Is the need to replace `_CHECKPOINT_WRAPPED_MODULE` a consequence of our usage of the `apply_activation_checkpointing` API? And if so, can we remove this by switching to the torch.utils.checkpoint API? (Obviously not a concern for this PR, I am more just curious)

Reply:
you are right. `apply_activation_checkpointing` is polluting fqns and I have to get clean fqn by replacing str `_CHECKPOINT_WRAPPED_MODULE` 

If we use `torch.utils.checkpoint`, we would not need `_CHECKPOINT_WRAPPED_MODULE`